sdk/dotnet/Pulumi.Automation/Runtime/LanguageRuntimeService.cs

```diff
@@ -23,10 +24,69 @@ public LanguageRuntimeService(CallerContext callerContext)
             this._callerContext = callerContext;
         }
 
-        public override Task<GetRequiredPluginsResponse> GetRequiredPlugins(GetRequiredPluginsRequest request, ServerCallContext context)
+        public override async Task<GetRequiredPluginsResponse> GetRequiredPlugins(GetRequiredPluginsRequest request, ServerCallContext context)
         {
             var response = new GetRequiredPluginsResponse();
-            return Task.FromResult(response);
+            var assembly = this._callerContext.Program.DiscoveryAssembly;
+
+            // if discovery assembly is null, assuming consumer installed plugins manually
+            if (assembly is null)
+                return response;
+
+            var pulumiAssemblyNames = assembly.GetReferencedAssemblies().Where(x => x.Name != null && x.Name.StartsWith("Pulumi.")).ToArray();
+            foreach (var assemblyName in pulumiAssemblyNames)
+            {
+                var pulumiAssembly = Assembly.Load(assemblyName);
+
+                // get version.txt
+                // this part would probably be less error prone if there was an assembly attribute available
+                // instead of using the embedded version.txt resource
+                // pulumiAssembly.GetCustomAttribute<ResourcePluginAttribute>();
+                var resources = pulumiAssembly.GetManifestResourceNames();
+                var versionResource = resources.FirstOrDefault(x => x.EndsWith("version.txt"));
+                if (versionResource is null)
+                {
+                    // no version.txt so is not resource plugin
+                    continue;
+                }
+
+                using var versionStream = pulumiAssembly.GetManifestResourceStream(versionResource);
+                if (versionStream is null)
+                {
+                    // we've already verified that version.txt should be present so throw here?
+                    throw new ApplicationException($"Unable to load version.txt from {assemblyName.FullName}");
+                }
+
+                var versionReader = new StreamReader(versionStream);
+                var versionText = await versionReader.ReadToEndAsync().ConfigureAwait(false);
+
+                // ToLower() everything after "Pulumi." to determine plugin name
+                var pluginName = assemblyName.Name!.Substring("Pulumi.".Length).ToLower();
```

Comment:
`ToLower` uses the casing conventions of the current culture. If we keep this implementation, we should use `ToLowerInvariant` instead:

```suggestion
                var pluginName = assemblyName.Name!.Substring("Pulumi.".Length).ToLowerInvariant();
```